File: clients/go/internal/capture/capture.go
// Package capture is the oracle's capture conversion (opendaisugi/hook.py):
// the captured sessions under a captures root listed, and one session
// turned into a journal trace with an envelope inferred from what it did.
package capture

import (
	"errors"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode/utf8"

	"daisugi-verify/internal/pmodel"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
	"daisugi-verify/internal/verify"
)

// ErrUnreadable is a capture this binary does not convert the way Python
// does. The command refuses before it writes anything.
var ErrUnreadable = errors.New("cannot read")

// InvalidPlan is a session whose plan ActionPlan refuses in a way this
// binary words exactly: a NaN or an infinity in a step's data
// (models.non_finite_error). Python's ValidationError is a ValueError, so
// auto-tend skips the session with its text.
type InvalidPlan struct{ Text string }

func (e *InvalidPlan) Error() string { return e.Text }

// Session is one row of list_sessions.
type Session struct {
	ID     string
	Calls  int
	LastAt any
	Path   string
}

func lines(path string) ([]string, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	if !utf8.Valid(raw) {
		return nil, fmt.Errorf("%w: %s is not UTF-8", ErrUnreadable, filepath.Base(path))
	}
	text := strings.ReplaceAll(strings.ReplaceAll(string(raw), "\r\n", "\n"), "\r", "\n")
	parts := strings.SplitAfter(text, "\n")
	var out []string
	for _, p := range parts {
		if p != "" {
			out = append(out, p)
		}
	}
	return out, nil
}

func number(v any) (*big.Rat, bool) {
	switch x := v.(type) {
	case pyjson.Int:
		n, ok := new(big.Int).SetString(x.Text, 10)
		if !ok {
			return nil, false
		}
		return new(big.Rat).SetInt(n), true
	case pyjson.Float:
		r := new(big.Rat)
		if r.SetFloat64(float64(x)) == nil {
			return nil, false
		}
		return r, true
	case bool:
		if x {
			return big.NewRat(1, 1), true
		}
		return new(big.Rat), true
	}
	return nil, false
}

// ListSessions is list_sessions(root): every *.jsonl with a readable first
// and last line, newest last_at first.
func ListSessions(root string) ([]Session, error) {
	st, err := os.Stat(root)
	if err != nil || !st.IsDir() {
		if err == nil || errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	names, err := filepath.Glob(filepath.Join(globEscape(root), "*.jsonl"))
	if err != nil {
		return nil, err
	}
	sort.Strings(names)
	var rows []Session
	for _, f := range names {
		if fi, err := os.Stat(f); err != nil || fi.IsDir() {
			continue
		}
		ls, err := lines(f)
		if err != nil {
			return nil, err
		}
		var first, last string
		calls := 0
		for _, l := range ls {
			if pystr.Strip(l) == "" {
				continue
			}
			if calls == 0 {
				first = l
			}
			last = l
			calls++
		}
		if calls == 0 {
			continue
		}
		fv, err1 := pyjson.LoadsPy(first, 900)
		lv, err2 := pyjson.LoadsPy(last, 900)
		if err1 != nil || err2 != nil {
			continue
		}
		fo, ok1 := fv.(*pyjson.Object)
		lo, ok2 := lv.(*pyjson.Object)
		if !ok1 || !ok2 {
			return nil, fmt.Errorf("%w: a line of %s is not a JSON object", ErrUnreadable, filepath.Base(f))
		}
		_ = fo
		lastAt, _ := lo.Get("captured_at")
		if lastAt != nil {
			if _, ok := number(lastAt); !ok {
				return nil, fmt.Errorf("%w: captured_at in %s is not a number", ErrUnreadable, filepath.Base(f))
			}
		}
		rows = append(rows, Session{ID: strings.TrimSuffix(filepath.Base(f), ".jsonl"), Calls: calls, LastAt: lastAt, Path: f})
	}
	key := func(v any) *big.Rat {
		if n, ok := number(v); ok && n.Sign() != 0 {
			return n
		}
		return new(big.Rat)
	}
	sort.SliceStable(rows, func(i, j int) bool { return key(rows[i].LastAt).Cmp(key(rows[j].LastAt)) > 0 })
	return rows, nil
}

func globEscape(s string) string {
	r := strings.NewReplacer("*", "[*]", "?", "[?]", "[", "[[]")
	return r.Replace(s)
}

// Records reads a session's records: each non-blank line's JSON, lines
// that do not parse skipped. A record whose shape the conversion would
// raise on is ErrUnreadable.
func Records(path string) ([]*pyjson.Object, error) {
	ls, err := lines(path)
	if err != nil {
		return nil, err
	}
	var out []*pyjson.Object
	for _, l := range ls {
		l = pystr.Strip(l)
		if l == "" {
			continue
		}
		v, derr := pyjson.LoadsPy(l, 900)
		if derr != nil {
			continue
		}
		o, ok := v.(*pyjson.Object)
		if !ok {
			return nil, fmt.Errorf("%w: a record of %s is not a JSON object", ErrUnreadable, filepath.Base(path))
		}
		st, has := o.Get("step_type")
		if _, isStr := st.(string); !has || !isStr {
			return nil, fmt.Errorf("%w: a record of %s has no step_type", ErrUnreadable, filepath.Base(path))
		}
		for _, k := range []string{"command", "path", "url", "mcp_server", "mcp_tool"} {
			if v, ok := o.Get(k); ok && pyjson.Truthy(v) {
				if _, isStr := v.(string); !isStr {
					return nil, fmt.Errorf("%w: %s in a record of %s is not text", ErrUnreadable, k, filepath.Base(path))
				}
			}
		}
		if v, ok := o.Get("arguments"); ok && pyjson.Truthy(v) {
			if _, isObj := v.(*pyjson.Object); !isObj {
				return nil, fmt.Errorf("%w: arguments in a record of %s is not an object", ErrUnreadable, filepath.Base(path))
			}
		}
		out = append(out, o)
	}
	return out, nil
}

func strOr(o *pyjson.Object, k string) string {
	if v, ok := o.Get(k); ok {
		if s, isStr := v.(string); isStr {
			return s
		}
	}
	return ""
}

// posixPath is pathlib.PurePosixPath's root and parts.
func posixPath(s string) (root string, parts []string) {
	switch {
	case strings.HasPrefix(s, "//") && !strings.HasPrefix(s, "///"):
		root = "//"
	case strings.HasPrefix(s, "/"):
		root = "/"
	}
	for _, p := range strings.Split(s, "/") {
		if p != "" && p != "." {
			parts = append(parts, p)
		}
	}
	return root, parts
}

func pathStr(root string, parts []string) string {
	s := root + strings.Join(parts, "/")
	if s == "" {
		return "."
	}
	return s
}

// globForPath is hook._glob_for_path.
func globForPath(path string) string {
	if path == "" {
		return "**"
	}
	root, parts := posixPath(path)
	parent := parts
	if len(parts) > 0 {
		parent = parts[:len(parts)-1]
	}
	if root != "" {
		p := "/"
		if len(parts) > 0 {
			p = pathStr(root, parent)
		}
		return strings.TrimRight(p, "/") + "/**"
	}
	p := pathStr(root, parent)
	if p == "" || p == "." {
		return "./**"
	}
	return "./" + strings.TrimRight(p, "/") + "/**"
}

const maxObserveDepth = 4

// observedEffects is hook._observed_effects.
func observedEffects(command string, decompose bool) (heads, reads, writes []string) {
	var collect func(cmd string, depth int)
	collect = func(cmd string, depth int) {
		if depth > maxObserveDepth {
			return
		}
		var simple []string
		ok := false
		if decompose {
			d := verify.DecomposeCommand(cmd)
			if d.OK {
				reads = append(reads, d.Reads...)
				writes = append(writes, d.Writes...)
				simple, ok = d.Commands, true
			}
		}
		if !ok {
			simple = []string{cmd}
		}
		for _, s := range simple {
			if h, found := verify.ExtractShellHead(pystr.Strip(s)); found {
				heads = append(heads, h)
			}
			if p, found := verify.ParseInterpreter(s); found && !p.Opaque {
				for _, inner := range p.InnerCommands {
					collect(inner, depth+1)
				}
			}
		}
	}
	collect(command, 0)
	return heads, reads, writes
}

var urlHost = regexp.MustCompile(`^https?://([^/]+)`)

var sanctionedWrite = map[string]bool{"/dev/null": true, "/dev/stdout": true, "/dev/stderr": true}
var sanctionedRead = map[string]bool{"/dev/null": true, "/dev/stdin": true}

func sortedAny(set map[string]bool) []any {
	keys := make([]string, 0, len(set))
	for k := range set {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	out := make([]any, len(keys))
	for i, k := range keys {
		out[i] = k
	}
	return out
}

// InferEnvelope is hook.infer_envelope: an envelope admitting every
// captured call, stakes low, with a fresh id.
func InferEnvelope(records []*pyjson.Object, task string, decompose bool) (*pyjson.Object, error) {
	heads, reads, writes := map[string]bool{}, map[string]bool{}, map[string]bool{}
	hosts, mcp := map[string]bool{}, map[string]bool{}
	network := false
	for _, r := range records {
		switch strOr(r, "step_type") {
		case "shell":
			cmd := pystr.Strip(strOr(r, "command"))
			if cmd == "" {
				continue
			}
			h, rs, ws := observedEffects(cmd, decompose)
			for _, x := range h {
				heads[x] = true
			}
			for _, p := range rs {
				if !sanctionedRead[p] {
					reads[globForPath(p)] = true
				}
			}
			for _, p := range ws {
				if !sanctionedWrite[p] {
					writes[globForPath(p)] = true
				}
			}
		case "file_read":
			reads[globForPath(strOr(r, "path"))] = true
		case "file_write":
			writes[globForPath(strOr(r, "path"))] = true
		case "network":
			network = true
			if m := urlHost.FindStringSubmatch(strOr(r, "url")); m != nil {
				hosts[pystr.Lower(m[1])] = true
			}
		case "mcp":
			server, tool := strOr(r, "mcp_server"), strOr(r, "mcp_tool")
			if server != "" && tool != "" && !(server == "opencode" && tool == "apply_patch") {
				mcp[server+"/"+tool] = true
			}
		}
	}
	perm := pyjson.NewObject().Set("shell", len(heads) > 0).Set("shell_allowlist", sortedAny(heads)).
		Set("shell_allow_decomposition", decompose).Set("mcp_allowlist", sortedAny(mcp)).
		Set("file_read", sortedAny(reads)).Set("file_write", sortedAny(writes)).Set("network", network).
		Set("network_hosts", sortedAny(hosts)).Set("max_execution_time_s", pyjson.Int{Text: "60"}).
		Set("max_output_size_mb", pyjson.Int{Text: "20"})
	env := pyjson.NewObject().Set("generated_by", "opendaisugi.hook.infer_envelope").Set("task", task).
		Set("permissions", perm).Set("stakes", "low")
	out, verr := pmodel.Envelope.ValidateObject(env, pmodel.Python)
	if verr != nil {
		return nil, fmt.Errorf("%w: %v", ErrUnreadable, verr)
	}
	return out, nil
}

// Plan is ActionPlan(source="hook-capture", task, _records_to_steps(...)).
func Plan(records []*pyjson.Object, task string) (*pyjson.Object, error) {
	steps := []any{}
	prev := ""
	for i, r := range records {
		sid := fmt.Sprintf("s%d", i)
		deps := []any{}
		if prev != "" {
			deps = append(deps, prev)
		}
		var st *pyjson.Object
		switch strOr(r, "step_type") {
		case "shell":
			st = pyjson.NewObject().Set("id", sid).Set("depends_on", deps).Set("type", "shell").Set("command", strOr(r, "command"))
		case "file_read":
			st = pyjson.NewObject().Set("id", sid).Set("depends_on", deps).Set("type", "file_read").Set("path", strOr(r, "path"))
		case "file_write":
			st = pyjson.NewObject().Set("id", sid).Set("depends_on", deps).Set("type", "file_write").
				Set("path", strOr(r, "path")).Set("content", "")
		case "network":
			st = pyjson.NewObject().Set("id", sid).Set("depends_on", deps).Set("type", "network").Set("url", strOr(r, "url"))
		case "mcp":
			args := pyjson.NewObject()
			if v, ok := r.Get("arguments"); ok && pyjson.Truthy(v) {
				args = v.(*pyjson.Object)
			}
			st = pyjson.NewObject().Set("id", sid).Set("depends_on", deps).Set("type", "mcp").
				Set("server", strOr(r, "mcp_server")).Set("tool", strOr(r, "mcp_tool")).Set("arguments", args)
		}
		if st != nil {
			steps = append(steps, st)
		}
		prev = sid
	}
	plan := pyjson.NewObject().Set("source", "hook-capture").Set("task", task).Set("steps", steps)
	out, verr := pmodel.ActionPlan.ValidateObject(plan, pmodel.Python)
	if verr != nil {
		finite := true
		for _, e := range verr.Errs {
			finite = finite && e.Type == "finite_number"
		}
		if finite {
			return nil, &InvalidPlan{verr.String()}
		}
		return nil, fmt.Errorf("%w: %v", ErrUnreadable, verr)
	}
	return out, nil
}

File: clients/rust/src/capture.rs
//! The oracle's capture conversion (`opendaisugi/hook.py`): the captured
//! sessions under a captures root listed, and one session turned into a
//! journal trace with an envelope inferred from what it did. The Go
//! client's `internal/capture` is the reference.

use std::collections::BTreeSet;

use crate::gate::py::text::{lower, strip};
use crate::gate::pyjson::{loads_py, Object, Value};
use crate::pathways::pmodel::{take_model, Id, Mode};
use crate::pathways::PwErr;
use crate::shell_decompose::ShellParser;

fn unreadable(why: impl Into<String>) -> PwErr {
    PwErr::Unreadable(why.into())
}

/// Why a session is not converted.
pub enum Skip {
    /// A plan ActionPlan refuses in a way this binary words exactly: a NaN
    /// or an infinity in a step's data (`models.non_finite_error`).
    /// Python's ValidationError is a ValueError, so auto-tend skips the
    /// session with its text.
    InvalidPlan(String),
}

/// One row of `list_sessions`.
#[derive(Clone)]
pub struct Session {
    pub id: String,
    pub calls: usize,
    pub last_at: Value,
    pub path: String,
}

fn base_name(p: &str) -> &str {
    p.rsplit('/').next().unwrap_or(p)
}

/// The file's lines, each with its line break, as `readlines` of a text
/// file gives them (universal newlines).
fn lines(path: &str) -> Result<Vec<String>, PwErr> {
    let raw = std::fs::read(path).map_err(PwErr::Io)?;
    let text = String::from_utf8(raw).map_err(|_| unreadable(format!("{} is not UTF-8", base_name(path))))?;
    let text = text.replace("\r\n", "\n").replace('\r', "\n");
    Ok(text.split_inclusive('\n').filter(|l| !l.is_empty()).map(|l| l.to_string()).collect())
}

/// A bool, an int or a finite float as a float, for ordering.
fn number(v: &Value) -> Option<f64> {
    match v {
        Value::Int(t) => t.parse::<f64>().ok(),
        Value::Float(f) if f.is_finite() => Some(*f),
        Value::Bool(b) => Some(*b as i64 as f64),
        _ => None,
    }
}

/// `list_sessions(root)`: every *.jsonl with a readable first and last
/// line, newest last_at first.
pub fn list_sessions(root: &str) -> Result<Vec<Session>, PwErr> {
    match std::fs::metadata(root) {
        Ok(m) if m.is_dir() => {}
        Ok(_) => return Ok(vec![]),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(vec![]),
        Err(e) => return Err(PwErr::Io(e)),
    }
    let mut names: Vec<String> = vec![];
    for e in std::fs::read_dir(root).map_err(PwErr::Io)? {
        let e = e.map_err(PwErr::Io)?;
        let name = e.file_name().to_string_lossy().into_owned();
        // pathlib's glob "*" takes a name that starts with a dot too.
        if name.ends_with(".jsonl") {
            names.push(format!("{}/{name}", root.trim_end_matches('/')));
        }
    }
    names.sort();
    let mut rows = vec![];
    for f in names {
        if std::fs::metadata(&f).map(|m| m.is_dir()).unwrap_or(true) {
            continue;
        }
        let ls = lines(&f)?;
        let nonblank: Vec<&String> = ls.iter().filter(|l| !strip(l).is_empty()).collect();
        if nonblank.is_empty() {
            continue;
        }
        let (first, last) = (nonblank[0], nonblank[nonblank.len() - 1]);
        let (fv, lv) = match (loads_py(first, 900), loads_py(last, 900)) {
            (Ok(a), Ok(b)) => (a, b),
            _ => continue,
        };
        let lo = match (fv, lv) {
            (Value::Obj(_), Value::Obj(lo)) => lo,
            _ => return Err(unreadable(format!("a line of {} is not a JSON object", base_name(&f)))),
        };
        let last_at = lo.get("captured_at").cloned().unwrap_or(Value::Null);
        if !last_at.is_null() && number(&last_at).is_none() {
            return Err(unreadable(format!("captured_at in {} is not a number", base_name(&f))));
        }
        let id = base_name(&f).strip_suffix(".jsonl").unwrap_or("").to_string();
        rows.push(Session { id, calls: nonblank.len(), last_at, path: f.clone() });
    }
    let key = |v: &Value| number(v).filter(|n| *n != 0.0).unwrap_or(0.0);
    rows.sort_by(|a, b| key(&b.last_at).partial_cmp(&key(&a.last_at)).unwrap_or(std::cmp::Ordering::Equal));
    Ok(rows)
}

/// A session's records: each non-blank line's JSON, lines that do not
/// parse skipped. A record whose shape the conversion would raise on is
/// refused.
pub fn records(path: &str) -> Result<Vec<Object>, PwErr> {
    let mut out = vec![];
    let name = base_name(path);
    for l in lines(path)? {
        let l = strip(&l);
        if l.is_empty() {
            continue;
        }
        let o = match loads_py(l, 900) {
            Ok(Value::Obj(o)) => o,
            Ok(_) => return Err(unreadable(format!("a record of {name} is not a JSON object"))),
            Err(_) => continue,
        };
        if !matches!(o.get("step_type"), Some(Value::Str(_))) {
            return Err(unreadable(format!("a record of {name} has no step_type")));
        }
        for k in ["command", "path", "url", "mcp_server", "mcp_tool"] {
            if let Some(v) = o.get(k) {
                if v.truthy() && !matches!(v, Value::Str(_)) {
                    return Err(unreadable(format!("{k} in a record of {name} is not text")));
                }
            }
        }
        if let Some(v) = o.get("arguments") {
            if v.truthy() && !matches!(v, Value::Obj(_)) {
                return Err(unreadable(format!("arguments in a record of {name} is not an object")));
            }
        }
        out.push(o);
    }
    Ok(out)
}

fn str_of<'a>(o: &'a Object, k: &str) -> &'a str {
    match o.get(k) {
        Some(Value::Str(s)) => s,
        _ => "",
    }
}

/// `pathlib.PurePosixPath`'s root and parts.
fn posix_path(s: &str) -> (&'static str, Vec<&str>) {
    let root = if s.starts_with("//") && !s.starts_with("///") {
        "//"
    } else if s.starts_with('/') {
        "/"
    } else {
        ""
    };
    (root, s.split('/').filter(|p| !p.is_empty() && *p != ".").collect())
}

fn path_str(root: &str, parts: &[&str]) -> String {
    let s = format!("{root}{}", parts.join("/"));
    if s.is_empty() {
        ".".into()
    } else {
        s
    }
}

/// `hook._glob_for_path`.
fn glob_for_path(path: &str) -> String {
    if path.is_empty() {
        return "**".into();
    }
    let (root, parts) = posix_path(path);
    let parent: &[&str] = if parts.is_empty() { &parts } else { &parts[..parts.len() - 1] };
    if !root.is_empty() {
        let p = if parts.is_empty() { "/".to_string() } else { path_str(root, parent) };
        return format!("{}/**", p.trim_end_matches('/'));
    }
    let p = path_str(root, parent);
    if p.is_empty() || p == "." {
        return "./**".into();
    }
    format!("./{}/**", p.trim_end_matches('/'))
}

const MAX_OBSERVE_DEPTH: usize = 4;

/// `hook._observed_effects`.
fn observed_effects(parser: &mut ShellParser, command: &str, decompose: bool) -> (Vec<String>, Vec<String>, Vec<String>) {
    let (mut heads, mut reads, mut writes) = (vec![], vec![], vec![]);
    fn collect(
        p: &mut ShellParser,
        cmd: &str,
        depth: usize,
        decompose: bool,
        out: &mut (&mut Vec<String>, &mut Vec<String>, &mut Vec<String>),
    ) {
        if depth > MAX_OBSERVE_DEPTH {
            return;
        }
        let mut simple = None;
        if decompose {
            let d = p.decompose(cmd);
            if d.ok {
                out.1.extend(d.reads);
                out.2.extend(d.writes);
                simple = Some(d.commands);
            }
        }
        let simple = simple.unwrap_or_else(|| vec![cmd.to_string()]);
        for s in simple {
            if let Some(h) = crate::verify::extract_shell_head(strip(&s)) {
                if !h.is_empty() {
                    out.0.push(h);
                }
            }
            if let Some(pl) = crate::interpreter_parse::parse_interpreter(&s) {
                if !pl.opaque {
                    for inner in &pl.inner_commands {
                        collect(p, inner, depth + 1, decompose, out);
                    }
                }
            }
        }
    }
    collect(parser, command, 0, decompose, &mut (&mut heads, &mut reads, &mut writes));
    (heads, reads, writes)
}

fn url_host(url: &str) -> Option<&str> {
    let rest = url.strip_prefix("http://").or_else(|| url.strip_prefix("https://"))?;
    let end = rest.find('/').unwrap_or(rest.len());
    (end > 0).then(|| &rest[..end])
}

fn sorted(set: BTreeSet<String>) -> Value {
    Value::List(set.into_iter().map(Value::Str).collect())
}

/// `hook.infer_envelope`: an envelope admitting every captured call,
/// stakes low, with a fresh id.
pub fn infer_envelope(parser: &mut ShellParser, recs: &[Object], task: &str, decompose: bool) -> Result<Object, PwErr> {
    let (mut heads, mut reads, mut writes, mut hosts, mut mcp) =
        (BTreeSet::new(), BTreeSet::new(), BTreeSet::new(), BTreeSet::new(), BTreeSet::new());
    let mut network = false;
    for r in recs {
        match str_of(r, "step_type") {
            "shell" => {
                let cmd = strip(str_of(r, "command"));
                if cmd.is_empty() {
                    continue;
                }
                let (h, rs, ws) = observed_effects(parser, cmd, decompose);
                heads.extend(h);
                for p in rs {
                    if p != "/dev/null" && p != "/dev/stdin" {
                        reads.insert(glob_for_path(&p));
                    }
                }
                for p in ws {
                    if !["/dev/null", "/dev/stdout", "/dev/stderr"].contains(&p.as_str()) {
                        writes.insert(glob_for_path(&p));
                    }
                }
            }
            "file_read" => {
                reads.insert(glob_for_path(str_of(r, "path")));
            }
            "file_write" => {
                writes.insert(glob_for_path(str_of(r, "path")));
            }
            "network" => {
                network = true;
                if let Some(h) = url_host(str_of(r, "url")) {
                    hosts.insert(lower(h));
                }
            }
            "mcp" => {
                let (server, tool) = (str_of(r, "mcp_server"), str_of(r, "mcp_tool"));
                if !server.is_empty() && !tool.is_empty() && !(server == "opencode" && tool == "apply_patch") {
                    mcp.insert(format!("{server}/{tool}"));
                }
            }
            _ => {}
        }
    }
    let mut perm = Object::new();
    perm.set("shell", !heads.is_empty());
    perm.set("shell_allowlist", sorted(heads));
    perm.set("shell_allow_decomposition", decompose);
    perm.set("mcp_allowlist", sorted(mcp));
    perm.set("file_read", sorted(reads));
    perm.set("file_write", sorted(writes));
    perm.set("network", network);
    perm.set("network_hosts", sorted(hosts));
    perm.set("max_execution_time_s", Value::Int("60".into()));
    perm.set("max_output_size_mb", Value::Int("20".into()));
    let mut env = Object::new();
    env.set("generated_by", "opendaisugi.hook.infer_envelope");
    env.set("task", task);
    env.set("permissions", perm);
    env.set("stakes", "low");
    match take_model(Id::Envelope, Value::Obj(env), Mode::Python) {
        Ok(Value::Obj(o)) => Ok(o),
        Ok(_) => Err(unreadable("the inferred envelope is not a mapping")),
        Err(e) => Err(unreadable(e.text())),
    }
}

/// `ActionPlan(source="hook-capture", task, _records_to_steps(...))`.
pub fn plan(recs: &[Object], task: &str) -> Result<Result<Object, Skip>, PwErr> {
    let mut steps = vec![];
    let mut prev = String::new();
    for (i, r) in recs.iter().enumerate() {
        let sid = format!("s{i}");
        let deps = if prev.is_empty() { vec![] } else { vec![Value::Str(prev.clone())] };
        let mut st = Object::new();
        st.set("id", sid.as_str());
        st.set("depends_on", Value::List(deps));
        let t = str_of(r, "step_type");
        let made = match t {
            "shell" => {
                st.set("type", "shell");
                st.set("command", str_of(r, "command"));
                true
            }
            "file_read" => {
                st.set("type", "file_read");
                st.set("path", str_of(r, "path"));
                true
            }
            "file_write" => {
                st.set("type", "file_write");
                st.set("path", str_of(r, "path"));
                st.set("content", "");
                true
            }
            "network" => {
                st.set("type", "network");
                st.set("url", str_of(r, "url"));
                true
            }
            "mcp" => {
                st.set("type", "mcp");
                st.set("server", str_of(r, "mcp_server"));
                st.set("tool", str_of(r, "mcp_tool"));
                let args = match r.get("arguments") {
                    Some(v @ Value::Obj(_)) if v.truthy() => v.clone(),
                    _ => Value::Obj(Object::new()),
                };
                st.set("arguments", args);
                true
            }
            _ => false,
        };
        if made {
            steps.push(Value::Obj(st));
        }
        prev = sid;
    }
    let mut p = Object::new();
    p.set("source", "hook-capture");
    p.set("task", task);
    p.set("steps", Value::List(steps));
    match take_model(Id::ActionPlan, Value::Obj(p), Mode::Python) {
        Ok(Value::Obj(o)) => Ok(Ok(o)),
        Ok(_) => Err(unreadable("the plan is not a mapping")),
        Err(e) => {
            if e.errs.iter().all(|x| x.typ == "finite_number") {
                return Ok(Err(Skip::InvalidPlan(e.text())));
            }
            Err(unreadable(e.text()))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn globs_for_paths_are_the_hooks() {
        assert_eq!(glob_for_path(""), "**");
        assert_eq!(glob_for_path("a.txt"), "./**");
        assert_eq!(glob_for_path("sub//dir/./b.txt"), "./sub/dir/**");
        assert_eq!(glob_for_path("/"), "/**");
        assert_eq!(glob_for_path("/work/src/a.py"), "/work/src/**");
        assert_eq!(glob_for_path("/x"), "/**");
        assert_eq!(glob_for_path("//x/y"), "//x/**");
    }
}
